CriticalMass/UpdateDataOperation.swift

```diff
@@ -0,0 +1,62 @@
+//
+//  UpdateDataOperation.swift
+//  CriticalMaps
+//
+//  Created by Илья Глущук on 13/10/2019.
+//  Copyright © 2019 Pokus Labs. All rights reserved.
+//
+
+import Foundation
+
+final class UpdateDataOperation: AsyncOperation {
+    private struct SendLocationPostBody: Encodable {
+        let device: String
+        let location: Location
+    }
+
+    private let locationProvider: LocationProvider
+    private let idProvider: IDProvider
+    private let networkLayer: NetworkLayer
+
+    var result: (Result<ApiResponse, NetworkError>)?
+
+    init(locationProvider: LocationProvider, idProvider: IDProvider, networkLayer: NetworkLayer) {
+        self.locationProvider = locationProvider
+        self.idProvider = idProvider
+        self.networkLayer = networkLayer
+        super.init()
+    }
+
+    override func main() {
+        guard let currentLocation = locationProvider.currentLocation else {
+            getLocationsAndChatMessages()
+            return
+        }
+        
+        let body = SendLocationPostBody(device: idProvider.id, location: currentLocation)
+        guard let bodyData = try? body.encoded() else {
+            state = .finished
+            return
+        }
+
+        let request = PostLocationRequest()
+        networkLayer.post(request: request, bodyData: bodyData) { [weak self] result in
+            guard let self = self else { return }
```

Comment:
Shouldn't we set state to finished?

Reply:
I set state to finished in completion block after result was assigned. Do you think we should set finished before assignment?

Reply:
If we are talking about "return" in guard, then self doesn't exists there. It means that operation queue was deallocated (with request manager) and I think we don't need to set state to finished there. By the way there is compilation error: "Variable declared in 'guard' condition is not usable in its body".

Reply:
oh makes sense 👍